### PR TITLE
LDAP authentication for GCC admin users

### PR DESCRIPTION
**Pull request by n1mh**
This pull request allows Gecos Control Center to authentica users via LDAP. It's mandatory to create some new variables in `gecoscc.ini` file for preventing unexpected errors. There's other pull request to change that file in gecoscc-installer repo.

### CODE REVIEW

File: gecoscc/userdb.py [hunk 1]
```diff
@@ -166,8 +186,52 @@ def get_user(request):
...
+        ld.unbind_s()
+        return False
+
+def get_ldap_userin(username):
+    ldap_info = get_ldap_info()
+    conn = str(ldap_info['url'])+':'+str(ldap_info['port'])
+    basedn = str(ldap_info['base'])
+    admn = str(ldap_info['admn'])
+    password = str(ldap_info['pass'])
+    sfilter = "(|(%s %s)(sAMAccountName=%s))" % (ldap_info['prfx'], username, username)
```
Comment (alfem) [inline gecoscc/userdb.py]: I am not sure every LDAP has got this attribute (sAMAccountName). I think is typical in AD, but not in LDAP.

File: gecoscc/utils.py [hunk 2]
```diff
@@ -944,21 +944,34 @@ def getURLComponents(url):
...
+
+    ldap__url = settings.get('gecos.ldap.url')
+    ldap_port = settings.get('gecos.ldap.port')
+    ldap_base = settings.get('gecos.ldap.base')
+    ldap_user = settings.get('gecos.ldap.user')
+    ldap_pass = settings.get('gecos.ldap.pass')
+
+    return ldap_user
+
 def update_computers_of_user(db, user, api):
```
Comment (n1mh) [inline gecoscc/utils.py]: this method was no longer used so I deleted it.

File: gecoscc/views/portal.py [hunk 3]
```diff
@@ -54,8 +54,30 @@ def login(self):
...
+            if not (user_authtype and user_authtype.has_key('authtype')):
+                self.request.userdb.add_authtype(username)
+                user_authtype = 'local'
+
+            ldap = get_ldap_info()
+            if ( ldap['url']  == None and user_authtype['authtype'] == 'ldap'):
+                return {
+                    'username': username,
+                    'message': self.translate(
+                        _("gecoscc.ini LDAP configuration misplaced. Please contact your system administrator.")),
```
Comment (n1mh) [inline gecoscc/views/portal.py]: changed for `wrong`.

File: config-templates/development.ini [hunk 4]
```diff
@@ -105,6 +109,17 @@ supervisord.url.pattern = http://internal:changeme@%s:9001/RPC2
 # This pattern could be HTTP or HTTPS depending on your configuration
 gecos.internal.url.pattern = http://%s/internal/server/%s
 
+# GecosCC LDAP configuration
+#
+# Change this variables in case you want to authentica users
+# using LDAP/AD directory
+gecos.ldap.url  = ldap://192.168.1.124
+gecos.ldap.port = 389 
+gecos.ldap.base = cn=users,dc=solutia-it,dc=es
```
Comment (n1mh) [inline config-templates/development.ini]: changed for `example.com`.

File: gecoscc/userdb.py (hunk 1 above)
Comment (n1mh) [inline gecoscc/userdb.py]: deleted and tested.

File: gecoscc/utils.py [hunk 5]
```diff
@@ -189,7 +189,7 @@ def _get_chef_api(chef_url, username, chef_pem, chef_ssl_verify, chef_version =
     return api
 
 
-def create_chef_admin_user(api, settings, usrname, password=None, email='nobody@nobody.es'):
+def create_chef_admin_user(api, settings, usrname, password=None, email='nobody@nobody.es', authtype='local'):
```
Comment (n1mh) [inline gecoscc/utils.py]: deleted.

File: gecoscc/utils.py [hunk 6]
```diff
@@ -945,20 +945,21 @@ def getURLComponents(url):
     return components
 
 def update_computers_of_user(db, user, api):
```
Comment (n1mh) [inline gecoscc/utils.py]: I have replace '\t' with whitespaces. Rolling back.

File: gecoscc/forms.py [hunk 7]
```diff
@@ -91,7 +91,7 @@ class GecosTwoColumnsForm(GecosForm):
 class BaseAdminUserForm(GecosTwoColumnsForm):
 
     sorted_fields = ('username', 'email', 'password',
-                     'repeat_password', 'first_name', 'last_name',
+                     'repeat_password', 'first_name', 'last_name', 'authtype',
```
Comment (n1mh) [inline gecoscc/forms.py]: authtype variable has been renamed to cc_auth_type

**Pull request closed without merging**